pyomo/contrib/sensitivity_toolbox/sens.py

```diff
@@ -218,7 +218,213 @@ def sipopt(instance, paramSubList, perturbList,
+    #      ComponentMap [ECSA 2018/11/23], this relates to Issue #755
+    paramCompMap = ComponentMap(zip(paramSubList, varSubList))
+    variableSubMap = {}
+    # variableSubMap = ComponentMap()
+    paramPerturbMap = ComponentMap(zip(paramSubList,perturbList))
+    perturbSubMap = {}
+    # perturbSubMap = ComponentMap()
+   
+    paramDataList = [] 
+    for parameter in paramSubList:
+        # Loop over each ParamData in the Param Component
+        #
+        # Note: Sets are unordered in Pyomo.  For this to be
+        # deterministic, we need to sort the index (otherwise, the
+        # ordering of things in the paramDataList may change).  We use
+        # sorted_robust to guard against mixed-type Sets in Python 3.x
+        for kk in sorted_robust(parameter):
+            variableSubMap[id(parameter[kk])]=paramCompMap[parameter][kk]
+            perturbSubMap[id(parameter[kk])]=paramPerturbMap[parameter][kk]
+            paramDataList.append(parameter[kk])
+
+    # clone Objective, add to Block, and update any Expressions
+    for cc in list(m.component_data_objects(Objective,
+                                            active=True,
+                                            descend_into=True)):
+        tempName=unique_component_name(m,cc.local_name)    
+        b.add_component(tempName,
+                  Objective(expr=ExpressionReplacementVisitor(
+                  substitute=variableSubMap,
+                  remove_named_expressions=True).dfs_postorder_stack(cc.expr)))
+        cc.deactivate()
+    
+    # clone Constraints, add to Block, and update any Expressions
+    b.constList = ConstraintList()
+    for cc in list(m.component_data_objects(Constraint, 
+                                   active=True,
+                                   descend_into=True)):
+        if cc.equality:
+            b.constList.add(expr= ExpressionReplacementVisitor(
+                    substitute=variableSubMap,
+                    remove_named_expressions=True).dfs_postorder_stack(cc.expr))
+        else:
+            raise Exception('kaug does not support inequalities. Need to replace inequalities to equalities with slack variables')
+        cc.deactivate()
+
+    # paramData to varData constraint list
+    b.paramConst = ConstraintList()
+    for ii in paramDataList:
+        jj=variableSubMap[id(ii)]
+        b.paramConst.add(ii==jj)
+
+    
+    # Create the solver plugin using the ASL interface
+    ipopt = SolverFactory('ipopt',solver_io='nl')
+    if optarg is not None:
+        ipopt.options = optarg
+    kaug = SolverFactory('k_aug',solver_io='nl')
+    dotsens = SolverFactory('dot_sens',solver_io='nl')
+    if not ipopt.available(False):
+        raise ImportError('ipopt is not available')
+    if not kaug.available(False):
+        raise ImportError('k_aug is not available')
+    if not dotsens.available(False):
+        raise ImportError('dotsens is not available')
```

Comment:
I know you were copying the structure of the sipopt implementation but I think these solver availability checks should be moved much earlier in the function. It doesn't make much sense to go through all the effort of modifying a large proportion of the model if the solvers aren't even available to finish the analysis.

Reply:
I agree with you. I moved them to the beginning of the function.